daemon/cmd/app.go

```diff
@@ -0,0 +1,224 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/spf13/cobra"
+	"go.uber.org/fx"
+
+	"github.com/cilium/cilium/api/v1/server"
+	"github.com/cilium/cilium/pkg/datapath/iptables"
+	linuxdatapath "github.com/cilium/cilium/pkg/datapath/linux"
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/endpoint"
+	"github.com/cilium/cilium/pkg/endpointmanager"
+	"github.com/cilium/cilium/pkg/ipcache"
+	"github.com/cilium/cilium/pkg/ipmasq"
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/mtu"
+	nodeTypes "github.com/cilium/cilium/pkg/node/types"
+	"github.com/cilium/cilium/pkg/option"
+	wireguard "github.com/cilium/cilium/pkg/wireguard/agent"
+)
+
+const (
+	startTimeout = 5 * time.Minute
+	stopTimeout  = 5 * time.Minute
+)
+
+var linuxDatapathModule = fx.Module(
+	"linux-datapath",
+
+	fx.Provide(
+		func(config *option.DaemonConfig) linuxdatapath.DatapathConfiguration {
+			return linuxdatapath.DatapathConfiguration{
+				HostDevice: defaults.HostDevice,
+				ProcFs:     config.ProcFs,
+			}
+		}),
+
+	fx.Provide(
+		linuxdatapath.NewDatapath,
+	),
+)
+
+var daemonModule = fx.Module(
+	"daemon",
+
+	fx.Provide(
+		NewDaemon,
+
+		// Lift out some objects from the daemon for modules.
+		func(d *Daemon) (*ipcache.IPCache, mtu.Configuration) {
+			return d.ipcache, d.mtuConfig
+		},
+	),
+
+	fx.Invoke(
+		preInit,
+		registerDaemonStart,
+		endpointManagerInit,
+	),
+)
+
+func runApp(cmd *cobra.Command) {
+	// Create a top-level context for the daemon. This is cancelled by the signal handler
+	// in cleanup.go and will trigger stop.
+	ctx, cancel := context.WithCancel(server.ServerCtx)
+
+	initEnv(cmd)
+
+	app := fx.New(
+		fx.StartTimeout(startTimeout),
+		fx.StopTimeout(stopTimeout),
+		fx.WithLogger(newAppLogger),
+
+		fx.Supply(
+			fx.Annotate(ctx, fx.As(new(context.Context))),
+			cancel),
+
+		fx.Supply(option.Config),
+
+		fx.Provide(
+			iptables.NewIptablesManager,
+			newEndpointManager,
+		),
+
+		// The order in which the modules are declared denotes the invoke and
+		// OnStart order.
+		linuxDatapathModule,
+		daemonModule,
+		wireguard.Module,
+		optional(option.Config.EnableIPMasqAgent, ipmasqAgentModule),
```

Comment:
Broadly the way this is being factored out seems reasonable enough. I think we will want to ensure that event generators like k8s or the Cilium API really are started pretty late to ensure that all of the other subsystems are up and ready to run / handle events first, to make sure we don't lose events or even crash (if there's some uninitialized object access or something that's not currently handled by the reworking into this DI framework).

I note in particular that WireGuard is started very early in the current init, but now this is being moved much later, and the IPMasqAgent logic which currently runs fairly late (just before agent API is exposed) is also later.